### ISSUE TITLE
API proposal: Add Insert extension method for IConfigurationBuilder

### ISSUE THREAD

**Issue by fredrikhr** [dotnet/runtime#38536]
## Background and Motivation

It is common for Configuration providers (such as `Microsoft.Extensions.Configuration.Json`) to provide extension methods that add configuration sources to a `IConfigurationBuilder`.

The .NET Generic Host utility method `Host.CreateDefaultBuilder` also adds configuration sources as part of the Generic Host infrastructure. However, there are cases where you'd like to keep all the defaults but add your own configuration sources **in front** of the defaults, for example by adding constant in-memory configurations or assembly-embedded configuration sources.

In order to avoid that all configuration source providers need to implement this functionality themselves, the new API should try to cope with the existing API structure.

## Proposed API

``` cs
#nullable enable
using System;

namespace Microsoft.Extensions.Configuration
{
    // Existing extension holder class
    public static partial class ConfigurationExtensions
    {
        // NEW proposed API
        public static IConfigurationBuilder Insert(
            this IConfigurationBuilder configBuilder, int index,
            Action<IConfigurationBuilder>? configAction) => null!;
    }
}
```

Propsed API would be added to the `Microsoft.Extensions.Configuration.Abstractions` package.

## Usage Examples

The following example shows the usage of the `Insert` method inside the `ConfigureAppConfiguration` method of the .NET Generic Host.

The example configures the configuration value `SampleAppSetting` coming from a static in-memory source. The `Host.CreateDefaultBuilder` adds additional configuration sources (like Command-line, Environment variables, `appsettings.json`, etc.). These additional sources would now all be able to override the static value `ConstantValue` defined here.

``` cs
using System.Collections.Generic;
using System.Threading.Tasks;

using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Configuration;

namespace Sample
{
    public static class Program
    {
        public static Task Main(string[] args)
        {
            return Host.CreateDefaultBuilder(args)
                .ConfigureAppConfiguration(config =>
                {
                    // Insert at 1, because host-configuration is placed at 0.
                    config.Insert(1, c =>
                    {
                        c.AddInMemoryCollection(new Dictionary<string, string>
                        {
                            ["SampleAppSetting"] = "ConstantValue",
                        });
                    });
                })
                .RunConsoleAsync();
        }
    }
}
```

If `Insert` was not used here, and `AddInMemoryCollection` was used directly, the in-memory collection would be added last, as thus be the defining source for the configuration value.

## Why a callback parameter?

The existing implementors for Configuration infrastructure usually implement their functionality by providing extension methods that all somehow end up calling `Add` on the `IConfigurationBuilder`. Examples of these are:

- `AddInMemoryCollection`
- `AddJsonFile`
- `AddIniFile`
- ...

While we also could add an `Insert` method that better matches the signature of `Add` there would be less value in doing so, since that would force all extensions to now also provide Insert extension methods for that to be useful.

Instead this proposal takes a callback instead, in which the user is expected to `Add` configuration sources (either by calling `Add` directly [uncommon], or by calling the one or multiple of the `Add`-extension methods listed above.

An implementation of the proposal would simply capture the count before invoking the callback, and then subsequently move all newly added configuration sources to the desired index.

## Risks

**None**

The proposed method will mutate the list of Configuration sources using API surface that is already publicly available.

~Using the `Insert` method could potentially break configuration providers that rely on a specific ordering of configuration sources. It would be the callers responsibility to ensure that configurations sources are added with a sensible value for the `index` parameter.~

~For example, one configuration provider (`provider-a`) could implement an `Add`-extension method that adds two configuration sources to the builder consecutively. Ad subsequent call to insert could insert itself between those two configuration source, potentially breaking the implementation of `provider-a` that relies on its two sources to directly follow each other in the list of sources. Such an example should be a very rare occurence, and in such cases the implementation of `provider-a` is non-ideal.~

**Comment by maryamariyan** [comment]
Thanks @fredrikhr. Before moving ahead with this API proposal I think is to explore the risks it would introduce on the configuration providers.

**Comment by terrajobst** [comment]
@maryamariyan request to table this for now so that they can think through the potential breaking changes first.

**Comment by menees** [comment]
I came across this issue while looking for an API to insert a new configuration source in front of the defaults. Ultimately, I just added a new source and used the mutable list returned by `IConfigurationBuilder.Sources` to move the last item to be first:

``` C#
// Inside ConfigureAppConfiguration:
byte[] jsonBytes = EmbeddedFile.ReadAllBytes<Startup>("appsettings.Embedded.json");
config.AddJsonStream(new MemoryStream(jsonBytes)); // The builder will close the stream on Build().
IList<IConfigurationSource> sources = config.Sources;
int removeIndex = sources.Count - 1;
IConfigurationSource embeddedJsonSource = sources[removeIndex];
sources.RemoveAt(removeIndex);
sources.Insert(0, embeddedJsonSource);
```

Since almost all IConfigurationBuilder extensions work in an Add (last) mode, an Insert extension goes a little against the grain. If anything it would probably be safer to add an extension API like `MoveSource(int fromIndex, int toIndex)`. Then moving the last source to first would just be `config.MoveSource(config.Sources.Count - 1, 0)`. You could fancy it up with other extensions like `MoveLastSourceFirst()` and possibly overloads to take Index instances instead of int (e.g., `MoveSource(^1, 0)`).

**Comment by fredrikhr** [comment]
@menees Well, I have a working implementation. And it indeed uses the mutable list in the way you describe. I first save the count of the list, then call Add, then infer the indicies to move (by comparing the new Count to the old count).

The intent of my proposal is that no IConfiguration implementor needs to change Add-logic and all the extensions that all already exist. The proposal here is to subsequently move the added sources to the desired position.

**Comment by menees** [comment]
Cool. Then this extension method doesn't _introduce_ any risk (using @maryamariyan's term). The existing, public mutable API could already break an order-dependent provider. That's a provider problem not an extension problem. Reviewing the existing providers should be independent of this extension proposal.

Plus, it already [works on my machine](https://blog.codinghorror.com/the-works-on-my-machine-certification-program/). ;-)

**Comment by fredrikhr** [comment]
@maryamariyan I agree with @menees here and have updated the `Risks` section of the proposal stating that this API would only use already existing publicly available API and thus does not introduce any new risk.

Unless someone has additional concerns, I'd suggest you want to re-review?
